### PR TITLE
Updates tutorial with entry on how to run Codex with its Markeplace

### PR DESCRIPTION
**Pull request by gmega**
This PR adds a first version of a guide for running Codex with a local Geth network and buying/selling storage in the Marketplace.

### CODE REVIEW

**Comment by gmega** [comment]
Ignore the CI failure as this is just updating text files. :-)

File: docs/Marketplace.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+Although the process is not particularly complicated, I suggest you use [the script we prepared](https://github.com/gmega/local-codex-bare/blob/main/scripts/mint-tokens.js) for that. This script, essentially:
+
+1. reads the Marketplace contract address and its ABI from the deployment data;
+2. transfers $1$ ETH from the signer account to a target account if the target account has no ETH balance;
+3. mints $n$ Codex tokens and adds it into the target account's balance.
+
+To use the script, just copy it locally into a file named `mint-tokens.js`, and run:
+
+```bash
+# Installs Web3-js
```
Comment (elcritch) [inline docs/Marketplace.md:171]:
```suggestion
To use the script, just download it into a local file named `mint-tokens.js` or use curl:

```bash
curl https://github.com/gmega/local-codex-bare/blob/main/scripts/mint-tokens.js -o mint-token.js
```

Then run the following to mine the tokens:
```bash
# Installs Web3-js
```

File: docs/Marketplace.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  --miner.etherbase <signer-account-address>\
+  --http\
+  --allow-insecure-unlock
+```
+
+Note that, once again, the signer account created in Step 1.1 appears both in `--unlock` and `--allow-insecure-unlock`: do not forget to insert it there. Geth will prompt you to insert the account's password as it starts up. Once you do that, it should be able to start up and begin "mining" blocks.
+
+## 2. Set Up The Marketplace
+
+Setting up the marketplace entails: _i)_ deploying the Codex Marketplace contracts to our private blockchain, and _ii)_ provisioning accounts we will later use to buying and selling storage in the marketplace with the required token balances.
```
Comment (elcritch) [inline docs/Marketplace.md:117]: Numbered bullets here are probably a bit easier to follow (for me at least). Also `buying/selling` don't match the rest of the sentence:

```suggestion
Setting up the Codex marketplace entails:

1. Deploying the Codex Marketplace contracts to our private blockchain
2. Setup Ethereum accounts we will use to buy and sell storage in the Codex marketplace 
3. Provisioning those accounts with the required token balances

```

File: docs/Marketplace.md [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+If the command completes successfully, you are ready to prepare the accounts.
+
+### 2.2. Generate the Required Accounts
+
+We will run $2$ Codex nodes: a **storage provider**, which will sell storage on the network, and a **client**, which will buy and use such storage; we therefore need two valid Ethereum accounts. We could create random accounts by using one of the many  tools available to that end but, since this is a tutorial running on a local private network, we will simply provide you with two pre-made accounts along with their private keys which you can copy and paste instead:
+
+**Storage:**
+```text
+address: 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37
+private key: 0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3
```
Comment (elcritch) [inline docs/Marketplace.md:150]:
```suggestion
```sh
ETH_STORAGE_ADDR=0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37
ETH_STORAGE_PK=0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3
```

File: docs/Marketplace.md [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+```
+**Client:**
+```text
+address: 0x9F0C62Fe60b22301751d6cDe1175526b9280b965
+private key: 0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc
```
Comment (elcritch) [inline docs/Marketplace.md:155]:
```suggestion
```sh
ETH_CLIENT_ADDR=0x9F0C62Fe60b22301751d6cDe1175526b9280b965
ETH_CLIENT_PK=0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc
```

File: docs/Marketplace.md [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+Repeat password:
+
+Your new key was generated
+
+Public address of the key:   0x93976895c4939d99837C8e0E1779787718EF8368
+...
+```
+
+In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address; write it down.
+
```
Comment (elcritch) [inline docs/Marketplace.md:49]:
```suggestion
In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address. Save it for later usage.

Additionally set it as an environment variable for later usage:

```sh
export GETH_SIGNER_ADDR=0x93976895c4939d99837C8e0E1779787718EF8368
```


```

File: docs/Marketplace.md [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+To use the script, just copy it locally into a file named `mint-tokens.js`, and run:
+
+```bash
+# Installs Web3-js
+npm install web3
+# Provides tokens to the storage account.
+node ./mint-tokens.js <signer account address> 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000
+# Provides tokens to the client account.
+node ./mint-tokens.js <signer account address> 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000
```
Comment (elcritch) [inline docs/Marketplace.md:176]:
```suggestion
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000
# Provides tokens to the client account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000
```

File: docs/Marketplace.md [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+Don't forget to replace `<signer account address>` with the address of the signer account you created in Step 1.1.
```
Comment (elcritch) [inline docs/Marketplace.md:179]:
```suggestion
Don't forget to set the variable `GETH_SIGNER_ADDR` with the address of the signer account you created in Step 1.1. You can manually replace `$GETH_SIGNER_ADDR` with the address as well.
```

File: docs/Marketplace.md (hunk 6 above)
Comment (elcritch) [inline docs/Marketplace.md:176]:
```suggestion
# set the contract file location
export CONTRACT_DEPLOY_FULL="codex-contracts-eth/deployments/codexdisttestnetwork"

# Provides tokens to the storage account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x45BC5ca0fbdD9F920Edd12B90908448C30F32a37 10000000000

# Provides tokens to the client account.
node ./mint-tokens.js $GETH_SIGNER_ADDR 0x9F0C62Fe60b22301751d6cDe1175526b9280b965 10000000000
```

File: docs/Marketplace.md [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+**Address of the Marketplace Contract.** The contract address can be found inside of the file `codex-contracts-eth/deployments/codexdisttestnetwork/Marketplace.json`:
+
+```bash
+> grep '"address":' Marketplace.json
+  "address": "0x8891732D890f5A7B7181fBc70F7482DE28a7B60f",
+```
+
+**Prover ceremony files.** The ceremony files are under the `codex-contracts-eth/verifier/networks/codexdisttestnetwork` subdirectory. There are three of them: `proof_main.r1cs`, `proof_main.zkey`, and `prooof_main.wasm`. We will need all of them to start the Codex storage node.
+
```
Comment (elcritch) [inline docs/Marketplace.md:202]:
```suggestion

```sh
export CONTRACT_DEPLOY_FULL="codex-contracts-eth/deployments/codexdisttestnetwork"
export CONTRACT_VERIFIER=codex-contracts-eth/verifier/networks/codexdisttestnetwork/
```

Comment (elcritch) [reply docs/Marketplace.md:202]: I don't know how to quote the backticks inside the backticks. 😢

File: docs/Marketplace.md [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+Recall you have clone the `codex-contracts-eth` repository in Step 2.1. All of the required files are in there.
+
+**Address of the Marketplace Contract.** The contract address can be found inside of the file `codex-contracts-eth/deployments/codexdisttestnetwork/Marketplace.json`:
+
+```bash
+> grep '"address":' Marketplace.json
+  "address": "0x8891732D890f5A7B7181fBc70F7482DE28a7B60f",
+```
+
```
Comment (elcritch) [inline docs/Marketplace.md:200]: Then run the following with the correct market place address:

```sh
export MARKETPLACE_ADDRESS=0x8891732D890f5A7B7181fBc70F7482DE28a7B60f
```

File: docs/Marketplace.md [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+	"tolerance": 1,
+	"collateral": "1000"
+}'
+```
+
+The parameters under `--data` say that:
+
+1. we want to purchase storage for our file for $20$ minutes (`"duration": "1200"`);
+2. we are willing to pay up to $1$ token per byte, per second (`"reward": "1"`);
+3. we demand `1000` tokens in collateral for this storage request. Without getting into too much detail, this means we will be compensated with `1000` tokens for each partition in our file that storage nodes fail to provide timely proofs for;
```
Comment (AuHau) [inline docs/Marketplace.md:320]: When slashing nobody is "compensated", instead the tokens are mainly burned. There is also small portion of the tokens kept as an incentive reward for the future repair mechanism in order to compensate the nodes for the extra work they will need to do for the repair.

With specifying the collateral the user can indicate "how much valuable" is the data for him. Where most probably for higher collateral (and hence for the nodes having higher stake = motivation in the hosting of the content), the user will have to pay a premium price.

**Review by AuHau — COMMENTED** [review]
Left some comments.

Honestly, I am not sure why you went with the route of setting up your own Geth PoA network. If you use the Hardhat setup, you get all the essentials (blockchain node, funded wallets, contract deployment, contract address specification, etc.) with just one command. I understand the value of testing in a "production environment," but Hardhat should be, in my humble opinion, good enough for regular users. I suppose it then becomes a question of who this guide is for, but if we are targeting "people coming to play around on local machine", I would recommend using Hardhat more than local Geth.

File: docs/Marketplace.md [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+
+```bash
+${CODEX_BINARY}\
+  --data-dir=./codex-storage\
+  --listen-addrs=/ip4/0.0.0.0/tcp/8080\
+  --api-port=8000\
+  --disc-port=8090\
+  persistence\
+  --eth-provider=http://localhost:8545\
+  --eth-private-key=<(echo -n "0x06c7ac11d4ee1d0ccb53811b71802fa92d40a5a174afad9f2cb44f93498322c3")\
```
Comment (elcritch) [inline docs/Marketplace.md:219]:
```suggestion
  --eth-private-key=./storage.pkey\
```

File: docs/Marketplace.md [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+```bash
+${CODEX_BINARY}\
+  --data-dir=./codex-client\
+  --listen-addrs=/ip4/0.0.0.0/tcp/8081\
+  --api-port=8001\
+  --disc-port=8091\
+  --bootstrap-node=<storage-node-spr>
+  persistence\
+  --eth-provider=http://localhost:8545\
+  --eth-private-key=<(echo -n "0x5538ec03c956cb9d0bee02a25b600b0225f1347da4071d0fd70c521fdc63c2fc")\
```
Comment (elcritch) [inline docs/Marketplace.md:255]:
```suggestion
  --eth-private-key=./client.pkey\
```

File: docs/Marketplace.md [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+  --validator-max-slots=1000\
+  prover\
+  --circom-r1cs=${PROVER_ASSETS}/proof_main.r1cs\
+  --circom-wasm=${PROVER_ASSETS}/proof_main.wasm\
+  --circom-zkey=${PROVER_ASSETS}/proof_main.zkey
+```
+replacing each `${VALUE}` variable by their respective contents. We then extract the Signed Peer Record (SPR) of the storage node so we can bootstrap the client node with it. To get the SPR, issue the following call:
+
+```bash
+curl 'http://localhost:8080/api/codex/v1/debug/info'
```
Comment (elcritch) [inline docs/Marketplace.md:231]:
```suggestion
curl 'http://localhost:8000/api/codex/v1/debug/info'
```

File: docs/Marketplace.md [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+Before we can buy storage, we must have some actual data to request storage for. Start by uploading a small file to your client node. On Linux you could, for instance, use `dd` to generate a $100KB$ file:
+
+```bash
+dd if=/dev/urandom of=./data.bin bs=100K count=1
+```
+
+but any small file will do. Assuming your file is named `data.bin`, you can upload it with:
+
+```bash
+curl "http://localhost:8001/api/codex/v1/data" --data-bin @data1.bin
```
Comment (elcritch) [inline docs/Marketplace.md:297]:
```suggestion
curl "http://localhost:8001/api/codex/v1/data" --data-bin ./data1.bin
```

File: docs/Marketplace.md [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+Once the upload completes, you should see a CID (e.g. `zDvZRwzm2mK7tvDzKScRLapqGdgNTLyyEBvx1TQY37J2CdWdS6Sj`) for the file printed to the terminal. Use that CID in the purchase request:
+
+```bash
+curl "http://localhost:8081/api/codex/v1/storage/request/<upload CID>"
```
Comment (elcritch) [inline docs/Marketplace.md:303]:
```suggestion
curl "http://localhost:8001/api/codex/v1/storage/request/<upload CID>"
```

Comment (elcritch) [reply docs/Marketplace.md:303]: I'm getting an error I think you ran into previously: `Unable to encode manifest, not enough blocks, ecK = 2, blocksCount = 1%`

File: docs/Marketplace.md [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+Public address of the key:   0x93976895c4939d99837C8e0E1779787718EF8368
+...
+```
+
+In this example, the public address of the signer account is `0x93976895c4939d99837C8e0E1779787718EF8368`. Yours will print a different address. Save it for later usage.
+
+Next set an environment variable for later usage:
+
+```sh
+export GETH_SIGNER_ADDR=0x0000000000000000000000000000000000000000
```
Comment (elcritch) [inline docs/Marketplace.md:53]:
```suggestion
export GETH_SIGNER_ADDR="0x0000000000000000000000000000000000000000"
```

File: docs/Marketplace.md [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+```
+
+which should print something like:
+```sh
+  "address": "0x8891732D890f5A7B7181fBc70F7482DE28a7B60f",
+```
+
+Then run the following with the correct market place address:
+```sh
+export MARKETPLACE_ADDRESS=0x0000000000000000000000000000000000000000
```
Comment (elcritch) [inline docs/Marketplace.md:243]:
```suggestion
export MARKETPLACE_ADDRESS="0x0000000000000000000000000000000000000000"
```

File: docs/Marketplace.md [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+  --marketplace-address=${MARKETPLACE_ADDRESS}
+```
+
+## 4. Buy and Sell Storage on the Marketplace
+
+Any storage negotiation has two sides: a buyer and a seller. Before we can actually request storage, therefore, we must first put some of it for sale.
+
+### 4.1 Sell Storage
+
+The following request will cause the storage node to put out $50\text{MB}$ of storage for sale for $1$ hour, at a price of $1$ Codex token per byte per second, while expressing that it's willing to take at most a $1000$ Codex token penalty for missing a storage proof.
```
[the remark this answers is not in the document]
Comment (gmega) [reply docs/Marketplace.md]: Thanks @AuHau, I've rephrased and added a footnote.

**Comment by gmega** [comment]
> Left some comments.
> 
> Honestly, I am not sure why you went with the route of setting up your own Geth PoA network. If you use the Hardhat setup, you get all the essentials (blockchain node, funded wallets, contract deployment, contract address specification, etc.) with just one command. I understand the value of testing in a "production environment," but Hardhat should be, in my humble opinion, good enough for regular users. I suppose it then becomes a question of who this guide is for, but if we are targeting "people coming to play around on local machine", I would recommend using Hardhat more than local Geth.

Can't say I disagree with this as I was asking myself that same question later on - if we really want a geth network, we can have a separate tutorial on "how to set up geth PoA network for Codex" or something.

I won't change the guide at this point as I want to be done with it and this works as it is, but I do agree with what you're saying.

File: docs/Marketplace.md (hunk 10 above)
Comment (gmega) [reply docs/Marketplace.md:389]: Thanks, did some slight rephrasing which together with the footnote above should make it clear enough (I hope).

**Comment by AuHau** [comment]
Soooooo, I am not so sure about this 😅 Honestly, I would avoid having a "mainstream guide" for Marketplace with having Geth PoA... I wouldn't mind merging this as "advanced guide" for people who are being curious (although not sure why they would be curious about waisting time with setting up PoA rather testing our features). Seeing that the `TwoClientsTest` does not have any more Marketplace sections, I would either add them there as well together with "Hardhat blockchain" so we could refer "beginners" to that guide or rework this to use Hardhat and leave the original Geth PoA setup at the bottom as section "for interested readers".

**Comment by gmega** [comment]
Don't disagree, just probably won't have the time to do major changes on this anytime soon.

**Comment by elcritch** [comment]
> Don't disagree, just probably won't have the time to do major changes on this anytime soon.

@gmega, IMHO, if the tutorial works we should merge it as it's a complete work unit. Then someone can create a new PR to do as @AuHau suggested.

**Comment by gmega** [comment]
You need to convince @AuHau not me. :smile:

**Comment by dryajov** [comment]
> Soooooo, I am not so sure about this 😅 Honestly, I would avoid having a "mainstream guide" for Marketplace with having Geth PoA... I wouldn't mind merging this as "advanced guide" for people who are being curious (although not sure why they would be curious about waisting time with setting up PoA rather testing our features). Seeing that the `TwoClientsTest` does not have any more Marketplace sections, I would either add them there as well together with "Hardhat blockchain" so we could refer "beginners" to that guide or rework this to use Hardhat and leave the original Geth PoA setup at the bottom as section "for interested readers".

I think having a full fledged network setup guide is required, I'm not sure what is the actual problem. From what I can see, what is being added shows how to do precisely that.

**Review by dryajov — APPROVED** [review]
LGTM

**Comment by gmega** [comment]
I can see @AuHau's point that having a guide with hardhat would be easier when the goal is just looking into how the marketplace works (i.e. no point in having geth in the middle). But I'd also merge this and defer splitting the guide to when we have the bandwidth, as an imperfect guide is better than no guide. :slightly_smiling_face:

**Review by dryajov — APPROVED** [review]

**Comment by AuHau** [comment]
Yeah, let's merge it and I will try to work in Hardhat approach later on

**Review by elcritch — APPROVED** [review]

**Pull request merged**